From eff36afb55e098fc24718a9a096c2f1bd4f1744c Mon Sep 17 00:00:00 2001
From: Nora Matthias Schiffer <neocturne@universe-factory.net>
Date: Tue, 18 Nov 2025 19:45:21 +0100
Subject: [PATCH 1/3] gluon-core: firewall: merge reject_input_on_wan() and
 add_cellular_wan()

No need to iterate over zones twice.

This also updates the code to always add the cellular interfaces to the
wan zone.
---
 .../lib/gluon/upgrade/300-firewall-rules      | 31 ++++++++-----------
 1 file changed, 13 insertions(+), 18 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules b/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
index 897078ed52..b937818ec2 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
@@ -1,33 +1,28 @@
 #!/usr/bin/lua
 
 local uci = require('simple-uci').cursor()
-local platform = require 'gluon.platform'
-
 
 local defaults = uci:get_first('firewall', 'defaults')
 uci:set('firewall', defaults, 'input', 'REJECT')
 
-
-local function reject_input_on_wan(zone)
-	if zone.name == 'wan' then
-		uci:set('firewall', zone['.name'], 'input', 'REJECT')
-		uci:set('firewall', zone['.name'], 'conntrack', true)
+local function configure_wan(zone)
+	if zone.name ~= 'wan' then
+		return
 	end
 
-	return true
-end
+	uci:set('firewall', zone['.name'], 'input', 'REJECT')
+	uci:set('firewall', zone['.name'], 'conntrack', true)
 
-local function add_cellular_wan(zone)
-	if zone.name == 'wan' then
-		uci:set('firewall', zone['.name'], 'network', {'wan', 'wan6', 'cellular_4', 'cellular_6'})
-	end
-end
-
-uci:foreach('firewall', 'zone', reject_input_on_wan)
+	uci:set('firewall', zone['.name'], 'network', {
+		'wan',
+		'wan6',
+		'cellular_4',
+		'cellular_6',
+	})
 
-if platform.is_cellular_device() then
-	uci:foreach('firewall', 'zone', add_cellular_wan)
+	return true
 end
+uci:foreach('firewall', 'zone', configure_wan)
 
 for _, zone in ipairs({'mesh', 'loc_client', 'wired_mesh'}) do
 	-- Other packages assign interfaces to these zones

From 4f371dd6070737b014d287852b583ec1c39e51a5 Mon Sep 17 00:00:00 2001
From: Nora Matthias Schiffer <neocturne@universe-factory.net>
Date: Wed, 19 Nov 2025 00:42:02 +0100
Subject: [PATCH 2/3] gluon-wan-dnsmasq: use interface.d interface list for
 hotplug script

Make sure the two lists are always in sync (and easier to extend in
packages).
---
 .../files/etc/hotplug.d/iface/50-gluon-wan-dnsmasq              | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package/gluon-wan-dnsmasq/files/etc/hotplug.d/iface/50-gluon-wan-dnsmasq b/package/gluon-wan-dnsmasq/files/etc/hotplug.d/iface/50-gluon-wan-dnsmasq
index 82b51dcfbb..b0fb63f111 100644
--- a/package/gluon-wan-dnsmasq/files/etc/hotplug.d/iface/50-gluon-wan-dnsmasq
+++ b/package/gluon-wan-dnsmasq/files/etc/hotplug.d/iface/50-gluon-wan-dnsmasq
@@ -1,3 +1,3 @@
-if [ "$INTERFACE" = 'wan' -o "$INTERFACE" = 'wan6' -o "$INTERFACE" = 'cellular_4' -o "$INTERFACE" = 'cellular_6' ]; then
+if grep -qxFe "$INTERFACE" /lib/gluon/wan-dnsmasq/interface.d/*; then
 	/lib/gluon/wan-dnsmasq/update.lua
 fi

From 054a0022c6156856ffc6ce9145b9865ee976d95a Mon Sep 17 00:00:00 2001
From: Nora Matthias Schiffer <neocturne@universe-factory.net>
Date: Wed, 19 Nov 2025 00:43:58 +0100
Subject: [PATCH 3/3] gluon-core, gluon-wan-dnsmasq: also consider 'cellular'
 an uplink interface

QMI interfaces will not always create a separate 'cellular_6' interface
for IPv6. Also accept configuration found directly on the 'cellular'
interface to fix IPv6 DNS in this case.

The firewall rules are adjusted for consistency, although here the issue
appears to be cosmetic, as the 'cellular_4' interface results in wwan0
rules to be added to ip6tables as well.
---
 package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules | 1 +
 .../lib/gluon/wan-dnsmasq/interface.d/050-gluon-wan-dnsmasq    | 3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules b/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
index b937818ec2..ec622e896d 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/300-firewall-rules
@@ -16,6 +16,7 @@ local function configure_wan(zone)
 	uci:set('firewall', zone['.name'], 'network', {
 		'wan',
 		'wan6',
+		'cellular',
 		'cellular_4',
 		'cellular_6',
 	})
diff --git a/package/gluon-wan-dnsmasq/files/lib/gluon/wan-dnsmasq/interface.d/050-gluon-wan-dnsmasq b/package/gluon-wan-dnsmasq/files/lib/gluon/wan-dnsmasq/interface.d/050-gluon-wan-dnsmasq
index 6a5af835f0..fff9a92f25 100644
--- a/package/gluon-wan-dnsmasq/files/lib/gluon/wan-dnsmasq/interface.d/050-gluon-wan-dnsmasq
+++ b/package/gluon-wan-dnsmasq/files/lib/gluon/wan-dnsmasq/interface.d/050-gluon-wan-dnsmasq
@@ -1,4 +1,5 @@
-wan6
 wan
+wan6
+cellular
 cellular_4
 cellular_6
